File: src/agent-app-node/src/aiGatewayClient.js
/**
 * Client for calling Azure OpenAI through the APIM AI Gateway.
 * All LLM calls are routed through APIM for governance.
 */
import axios from 'axios';
import { getTracer } from './telemetry.js';
import { config } from './config.js';

const tracer = getTracer('macu-ai-gateway-client');

/**
 * Calls Azure OpenAI via the APIM AI Gateway.
 */
export class AIGatewayClient {
  constructor() {
    this.baseUrl = config.apimGatewayUrl.replace(/\/$/, '');
    this.headers = {
      'Content-Type': 'application/json',
      'Ocp-Apim-Subscription-Key': config.apimSubscriptionKey,
    };
  }

  /**
   * Generate an embedding vector via APIM → Azure OpenAI.
   * @param {string} text - Text to embed
   * @returns {Promise<number[]>} Embedding vector
   */
  async getEmbedding(text) {
    const span = tracer.startSpan('ai-gateway.embeddings');

    try {
      span.setAttribute('model', config.embeddingModel);
      span.setAttribute('input.length', text.length);

      const url = `${this.baseUrl}/openai/deployments/${config.embeddingModel}/embeddings?api-version=${config.apiVersion}`;

      const response = await axios.post(
        url,
        { input: text },
        { headers: this.headers, timeout: 30000 }
      );

      const embedding = response.data.data[0].embedding;

      span.setAttribute('embedding.dimensions', embedding.length);
      span.setAttribute('gateway.trace_id', response.headers['x-trace-id'] || '');
      span.setAttribute('gateway.region', response.headers['x-backend-region'] || '');

      return embedding;
    } catch (error) {
      span.recordException(error);
      throw error;
    } finally {
      span.end();
    }
  }

  /**
   * Call chat completion via APIM → Azure OpenAI.
   * @param {Array<{role: string, content: string}>} messages - Chat messages
   * @param {number} maxTokens - Max tokens for completion
   * @returns {Promise<object>} Chat completion response
   */
  async chatCompletion(messages, maxTokens = 1000) {
    const span = tracer.startSpan('ai-gateway.chat-completion');

    try {
      span.setAttribute('model', config.chatModel);
      span.setAttribute('max_tokens', maxTokens);
      span.setAttribute('messages.count', messages.length);

      const url = `${this.baseUrl}/openai/deployments/${config.chatModel}/chat/completions?api-version=${config.apiVersion}`;

      const response = await axios.post(
        url,
        { messages, max_tokens: maxTokens },
        { headers: this.headers, timeout: 60000 }
      );

      const result = response.data;
      const usage = result.usage || {};

      span.setAttribute('tokens.prompt', usage.prompt_tokens || 0);
      span.setAttribute('tokens.completion', usage.completion_tokens || 0);
      span.setAttribute('tokens.total', usage.total_tokens || 0);
      span.setAttribute('gateway.trace_id', response.headers['x-trace-id'] || '');
      span.setAttribute('gateway.region', response.headers['x-backend-region'] || '');
      span.setAttribute('gateway.tokens_remaining', response.headers['x-tokens-remaining'] || '');

      return result;
    } catch (error) {
      span.recordException(error);
      throw error;
    } finally {
      span.end();
    }
  }
}

File: src/agent-app/ai_gateway_client.py
"""
Client for calling Azure OpenAI through the APIM AI Gateway.
All LLM calls are routed through APIM for governance.
"""
import httpx
from opentelemetry import trace
from config import Config

tracer = trace.get_tracer("macu-ai-gateway-client")


class AIGatewayClient:
    """Calls Azure OpenAI via the APIM AI Gateway."""

    def __init__(self):
        self.base_url = Config.APIM_GATEWAY_URL.rstrip("/")
        self.headers = {
            "Content-Type": "application/json",
            "Ocp-Apim-Subscription-Key": Config.APIM_SUBSCRIPTION_KEY,
        }

    async def get_embedding(self, text: str) -> list[float]:
        """Generate an embedding vector via APIM → Azure OpenAI."""
        with tracer.start_as_current_span("ai-gateway.embeddings") as span:
            span.set_attribute("model", Config.EMBEDDING_MODEL)
            span.set_attribute("input.length", len(text))

            url = (
                f"{self.base_url}/openai/deployments/{Config.EMBEDDING_MODEL}"
                f"/embeddings?api-version={Config.API_VERSION}"
            )

            async with httpx.AsyncClient() as client:
                response = await client.post(
                    url,
                    headers=self.headers,
                    json={"input": text},
                    timeout=30.0,
                )
                response.raise_for_status()

            result = response.json()
            embedding = result["data"][0]["embedding"]

            span.set_attribute("embedding.dimensions", len(embedding))
            # Capture gateway headers for observability
            span.set_attribute("gateway.trace_id", response.headers.get("x-trace-id", ""))
            span.set_attribute("gateway.region", response.headers.get("x-backend-region", ""))

            return embedding

    async def chat_completion(self, messages: list[dict], max_tokens: int = 1000) -> dict:
        """Call chat completion via APIM → Azure OpenAI."""
        with tracer.start_as_current_span("ai-gateway.chat-completion") as span:
            span.set_attribute("model", Config.CHAT_MODEL)
            span.set_attribute("max_tokens", max_tokens)
            span.set_attribute("messages.count", len(messages))

            url = (
                f"{self.base_url}/openai/deployments/{Config.CHAT_MODEL}"
                f"/chat/completions?api-version={Config.API_VERSION}"
            )

            async with httpx.AsyncClient() as client:
                response = await client.post(
                    url,
                    headers=self.headers,
                    json={"messages": messages, "max_tokens": max_tokens},
                    timeout=60.0,
                )
                response.raise_for_status()

            result = response.json()
            usage = result.get("usage", {})

            span.set_attribute("tokens.prompt", usage.get("prompt_tokens", 0))
            span.set_attribute("tokens.completion", usage.get("completion_tokens", 0))
            span.set_attribute("tokens.total", usage.get("total_tokens", 0))
            span.set_attribute("gateway.trace_id", response.headers.get("x-trace-id", ""))
            span.set_attribute("gateway.region", response.headers.get("x-backend-region", ""))
            span.set_attribute("gateway.tokens_remaining",
                               response.headers.get("x-tokens-remaining", ""))

            return result